### PR TITLE
Replace thread_local with generator resume arguments in box_region.

### LINKED ISSUE #68922: Rewrite `box_region.rs` to use generator resume arguments
[`box_region.rs`](https://github.com/rust-lang/rust/blob/3d8778d767f0dde6fe2bc9459f21ead8e124d8cb/src/librustc_data_structures/box_region.rs) makes use of generators, and passes arguments into them via thread-local storage. Now that https://github.com/rust-lang/rust/pull/68524 has landed, this can be rewritten to use generator resume arguments instead, which is much cleaner and possibly also faster.

### PR DESCRIPTION
**Pull request by gizmondo**
Fixes #68922. 

Continuation of #70622. Added a short doc, hope it makes sense. 

r? @jonas-schievink

### CODE REVIEW

File: src/librustc_data_structures/box_region.rs [hunk 1]
```diff
@@ -1,4 +1,15 @@
-use std::cell::Cell;
+//! This module provides a way to deal with self-referential data.
+//!
+//! The main idea is to allocate such data in a generator frame and then
+//! give access to it by executing user-provided closures inside that generator.
+//! The module provides a safe abstraction for the latter task.
+//!
+//! The interface consists of two exported macros meant to be used together:
+//! * `declare_box_region_type` wraps a generator inside a struct with `access`
+//! method which accepts closures.
```
Comment (jonas-schievink) [inline src/librustc_data_structures/box_region.rs:9]:
```suggestion
//!   method which accepts closures.
```

File: src/librustc_data_structures/box_region.rs [hunk 2]
```diff
@@ -1,4 +1,15 @@
...
+//! * `box_region_allow_access` is a helper which should be called inside
+//! a generator to actually execute those closures.
```
Comment (jonas-schievink) [inline src/librustc_data_structures/box_region.rs:11]:
```suggestion
//!   a generator to actually execute those closures.
```

**Review by jonas-schievink — COMMENTED** [review]
Nice! I only have a couple of formatting nits.

**Comment by gizmondo** [comment]
> Nice! I only have a couple of formatting nits.

Ah sure, thanks.

**Pull request merged**